### PR TITLE
Do not ICE when matching an uninhabited enum's field

### LINKED ISSUE #69191: beta regression: ICE on Tried to access field 0 of union Layout
- [`rotor v0.6.3`](https://crater-reports.s3.amazonaws.com/beta-1.42-1/beta-2020-02-05/reg/rotor-0.6.3/log.txt)
- [`rotor-dns v0.4.0`](https://crater-reports.s3.amazonaws.com/beta-1.42-1/beta-2020-02-05/reg/rotor-dns-0.4.0/log.txt)
- [`rotor-tools v0.3.2`](https://crater-reports.s3.amazonaws.com/beta-1.42-1/beta-2020-02-05/reg/rotor-tools-0.3.2/log.txt)

cc @oli-obk 

I suspect (but have not verified) that this would be fixed by a backport of https://github.com/rust-lang/rust/pull/64987.

### PR DESCRIPTION
**Pull request by pnkfelix**
Fix #69191

### CODE REVIEW

**Comment by Centril** [comment]
cc @RalfJung

**Comment by pnkfelix** [comment]
The reason the test is written in that manner is that I had played with a few other methods of fixing this bug; one in particular messed around with the logic here: 

https://github.com/rust-lang/rust/blob/b818ccc74c993d41286731f67ce1e2bc2a3f1918/src/librustc/ty/layout.rs#L781-L795

where I was trying to replace the `def.is_enum()` logic in the match with something that instead checked `!is_struct` from the boolean defined below that match.

Long story short, that approach didn't work; but when I was working on it, I was worried enough about the handling of `#[repr(C)]` and `#[repr(<int-type>)]` that I wanted explicit test variants that checked those cases.

**Comment by pnkfelix** [comment]
Either @oli-obk or @eddyb would be suitable candidates for reviewing this.

**Comment by pnkfelix** [comment]
(also, the reason that the code doesn't just unconditionally `throw_ub!` even for fields that are within range is that doing so would break [a regression test](https://github.com/rust-lang/rust/blob/master/src/test/ui/consts/issue-64506.rs) from #64506.)

I'm sure there is a better solution to this problem. I laid out three different choices [on zulip](https://zulip-archive.rust-lang.org/131828tcompiler/47306ICEaccessfieldofuninhabitedenum69191.html#189839856), but this is the one I felt most confident about being good enough to backport to beta in the short term. (Because all it does it injects a `throw_ub!` invocation immediately before something that would otherwise cause the compiler to ICE.)

File: src/librustc_mir/interpret/place.rs [hunk 1]
```diff
@@ -410,6 +410,12 @@ where
                 stride * field
             }
             layout::FieldPlacement::Union(count) => {
+                // This is a narrow bug-fix for rust-lang/rust#69191: if we are
+                // trying to access absent field of uninhabited variant, then
+                // signal UB (but don't ICE the compiler).
+                if field >= count as u64 && base.layout.abi == layout::Abi::Uninhabited {
+                    throw_ub!(Unreachable);
```
Comment (RalfJung) [inline src/librustc_mir/interpret/place.rs]: This is quite the hack... Miri really shouldn't need special checking for `layout::Abi::Uninhabited`. The right behavior for the uninhabited type ought to just fall out of the correct behavior for all kinds of types.

@eddyb also seems like the hack of using `Union` layout for uninhabited enums is having some bad repercussions here? But generally I am wondering how this can ever be reached; creating an uninhabited enum shiuld already be UB earlier...

@pnkfelix to make this hack slightly less hacky, is it sufficient to only do this when `base.layout.ty.is_enum()`?

Comment (oli-obk) [reply src/librustc_mir/interpret/place.rs]: This is reached from const prop...

Comment (oli-obk) [reply src/librustc_mir/interpret/place.rs]: I think we should just merge this and figure out how to do it correctly after it's merged, so we get rid of the beta regression immediately

Comment (RalfJung) [reply src/librustc_mir/interpret/place.rs]: Oh no, this is const-prop again. :(
From [the backtrace](https://github.com/rust-lang/rust/issues/69191#issuecomment-587952118):

```
  14: rustc_mir::interpret::operand::<impl rustc_mir::interpret::eval_context::InterpCx<M>>::operand_field
  15: rustc_mir::interpret::operand::<impl rustc_mir::interpret::eval_context::InterpCx<M>>::eval_place_to_op
  16: rustc_mir::interpret::operand::<impl rustc_mir::interpret::eval_context::InterpCx<M>>::eval_operand
  17: rustc_mir::interpret::step::<impl rustc_mir::interpret::eval_context::InterpCx<M>>::eval_rvalue_into_place
  18: <rustc_mir::transform::const_prop::ConstPropagator as rustc::mir::visit::MutVisitor>::visit_statement
  19: <rustc_mir::transform::const_prop::ConstProp as rustc_mir::transform::MirPass>::run_pass
```

So likely const-prop is running some unreachable code, and rather unsurprisingly that is tripping other parts of the compiler. Cc @wesleywiser 

Here's the pre-regression MIR:
```
    bb0: {
        StorageLive(_1);                 // bb0[0]: scope 0 at src/main.rs:6:9: 6:14
        StorageLive(_2);                 // bb0[1]: scope 0 at src/main.rs:6:40: 6:44
        _2 = const Scalar(<ZST>) : std::option::Option<UninhabitedUnivariant>; // bb0[2]: scope 0 at src/main.rs:6:40: 6:44
                                         // ty::Const
                                         // + ty: std::option::Option<UninhabitedUnivariant>
                                         // + val: Value(Scalar(<ZST>))
                                         // mir::Constant
                                         // + span: src/main.rs:6:40: 6:44
                                         // + literal: Const { ty: std::option::Option<UninhabitedUnivariant>, val: Value(Scalar(<ZST>)) }
        _1 = const std::option::Option::<UninhabitedUnivariant>::unwrap(move _2) -> bb1; // bb0[3]: scope 0 at src/main.rs:6:40: 6:53
                                         // ty::Const
                                         // + ty: fn(std::option::Option<UninhabitedUnivariant>) -> UninhabitedUnivariant {std::option::Option::<UninhabitedUnivariant>::unwrap}
                                         // + val: Value(Scalar(<ZST>))
                                         // mir::Constant
                                         // + span: src/main.rs:6:45: 6:51
                                         // + literal: Const { ty: fn(std::option::Option<UninhabitedUnivariant>) -> UninhabitedUnivariant {std::option::Option::<UninhabitedUnivariant>::unwrap}, val: Value(Scalar(<ZST>)) }
    }

    bb1: {
        StorageDead(_2);                 // bb1[0]: scope 0 at src/main.rs:6:52: 6:53
        StorageLive(_3);                 // bb1[1]: scope 1 at src/main.rs:8:41: 8:43
        _3 = move ((_1 as _Variant).0: Void); // bb1[2]: scope 1 at src/main.rs:8:41: 8:43
        StorageDead(_3);                 // bb1[3]: scope 1 at src/main.rs:9:5: 9:6
        StorageDead(_1);                 // bb1[4]: scope 0 at src/main.rs:10:1: 10:2
        return;                          // bb1[5]: scope 0 at src/main.rs:10:2: 10:2
    }
```
In `bb1`, mir building attempt to access `(_1 as _Variant).0`, and that variant has `Union` layout, and thus the ICE.

I feel like there is something incoherent in the compiler: either layout generation should actually add an appropriate amount of fields to these variants, or MIR building should not emit field accesses to fields that do not exist in the layout. But it seems like a reaosnable invariant to expect that any field mentioned statically *anywhere* in the code, is also actually existing in the layout.

@eddyb is it possible to, at leats, make the layout of such dead variants actually have the right number of fields (as declared in the source)?

**Comment by RalfJung** [comment]
> bors r+

FWIW, I don't think this is the right fix. At the very least we should add a comment saying `FIXME temporary hack to work around incoherence between layout computation and MIR building`.

**Comment by oli-obk** [comment]
It is not the correct fix, I agree, but this is getting into stable soon. I am opening a workspace locally later today to start working on a real fix

File: src/librustc_mir/interpret/place.rs (hunk 1 above)
Comment (RalfJung) [reply src/librustc_mir/interpret/place.rs]: All right, I'll open an issue then about the layout/MIR incoherence.

Comment (oli-obk) [reply src/librustc_mir/interpret/place.rs]: :upside_down_face: we can also solve the ICE by moving https://github.com/rust-lang/rust/blob/master/src/librustc_mir/interpret/operand.rs#L368-L371 to the beginning of the function, thus avoiding the problematic situation entirely

Comment (RalfJung) [reply src/librustc_mir/interpret/place.rs]: Can we really? `op.layout.field(self, field)` will likely fail because `field` is not a valid field index.

Comment (RalfJung) [reply src/librustc_mir/interpret/place.rs]: I opened an issue for further discussion: https://github.com/rust-lang/rust/issues/69763

Comment (eddyb) [reply src/librustc_mir/interpret/place.rs]: > @eddyb is it possible to, at least, make the layout of such dead variants actually have the right number of fields (as declared in the source)?

They do, I even showed you the code and brought up that `Result<!, !>` has `Union(1)` for both variants days ago.

We fixed this months ago. I don't understand why this is a problem again now.

**Comment by eddyb** [comment]
@oli-obk Alright, but when you do, please include a plain revert commit of this PR.
(minus the test, I suppose)

Also, just so it doesn't get lost, I pointed out in https://github.com/rust-lang/rust/pull/69753#discussion_r388798445 that this problem shouldn't be possible nowadays, so this is some bizarre intersection of multiple edge cases.

**Comment by pnkfelix** [comment]
@bors r=oli

This PR (#69753) is more suitable than PR  #69768 for a beta backport, so I am re-approving it for landing.

I still do want PR #69768 to be the answer on nightly.
 * after PR #69768 (or something even more general to resolve issue #69763) lands, we should even remove the workaround that was added by this PR #69753

**Comment by pnkfelix** [comment]
Also, assuming this does end up in nightly, its fine for it to be part of a rollup.

**Comment by nikomatsakis** [comment]
Discussed in today's design meeting, decided to accept this change for beta backport.
